### PR TITLE
Feature/#21 kotlinx

### LINKED ISSUE #21: JSON Serialisation Library
Should we use Jackson or try a different one?

### PR DESCRIPTION
**Pull request by aanciaes**
Closes #21

### CODE REVIEW

File: src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+
+object JsonSettings {
+
+    val mapper = Json {
+        ignoreUnknownKeys = true
+        prettyPrint = true
+    }
+
```
Comment (monxalo) [inline src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt:13]: Maybe move these two helper functions into the test scope since these are only used there.

File: src/main/kotlin/test/ktortemplate/core/httphandler/DefaultRoutes.kt [hunk 2]
```diff
@@ -27,11 +28,13 @@ fun Route.defaultRoutes() {
         val totalElements = carService.count(pageRequest)
         val data = carService.list(pageRequest)
         call.respond(
-            PageResponse.from(
-                pageRequest = pageRequest,
-                totalElements = totalElements,
-                data = data,
-                path = call.request.path()
+            JsonSettings.toJson(
```
Comment (monxalo) [inline src/main/kotlin/test/ktortemplate/core/httphandler/DefaultRoutes.kt:31]: It this call to `.toJson()` required? `PageResponse` is already `@Serializable`, why is this needed here?

File: src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt (hunk 1 above)
Comment (felix19350) [reply src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt:19]: I think @aanciaes  did the right thing here, as this ensures that we use a consistent set of mapper options for our enoding/decoding purposes.

File: src/test/kotlin/test/ktortemplate/core/model/TestModels.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
+@file:UseSerializers(InstantAsStringSerializer::class)
```
Comment (felix19350) [inline src/test/kotlin/test/ktortemplate/core/model/TestModels.kt:1]: What does this do? It registers the `InstantAsStringSerializer` as the default serializer for instants?

File: src/test/kotlin/test/ktortemplate/core/model/TestModels.kt [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+object InstantAsStringSerializer : KSerializer<Instant> {
+    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("Instant", PrimitiveKind.STRING)
+    override fun serialize(encoder: Encoder, value: Instant) =
+        encoder.encodeString(DateTimeFormatter.ISO_INSTANT.format(value))
+
+    override fun deserialize(decoder: Decoder): Instant = Instant.parse(decoder.decodeString())
+}
+
+@Serializable
+data class TestInstantLongSerialization(@Serializable(with = InstantAsLongSerializer::class) val date: Instant)
```
Comment (felix19350) [inline src/test/kotlin/test/ktortemplate/core/model/TestModels.kt:28]: @aanciaes  this is an example of overriding a serializer for a particular class?

**Review by felix19350 — COMMENTED** [review]
I kinda like the idea of using kotlinx serialization instead of jackson TBH, in most cases we just need the @Serializable annotation and it handles things like sealed classes well.

In terms of fat jar size, is there any noticeable difference between this and jackson?

File: src/test/kotlin/test/ktortemplate/core/model/TestModels.kt (hunk 3 above)
Comment (monxalo) [reply src/test/kotlin/test/ktortemplate/core/model/TestModels.kt:1]: Yes, only in the context of this file. It saves trouble for adding `@Serializable(with = InstantAsStringSerializer::class)` for each fields.

File: src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt [hunk 5]
```diff
@@ -15,6 +19,9 @@ object JsonSettings {
         coerceInputValues = false
         classDiscriminator = "type"
         allowSpecialFloatingPointValues = false
+        serializersModule = SerializersModule {
+            contextual(PageResponse.serializer(Car.serializer()))
```
Comment (felix19350) [inline src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt:23]: @monxalo this forces us to update the core json settings with the specific types we want to use with the PageResponse (or any other generic class)?

**Review by felix19350 — COMMENTED** [review]
@monxalo I'm not sure about the advantages of the change you introduced. Check my comment in the context of the code please.

File: src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt (hunk 5 above)
Comment (monxalo) [reply src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt:23]: Not necessarily, `SerializersModule` has a `plus()` which can add more modules. This can be setup using dependency injection where each module can provide a `SerializersModule` that contributes to `serializersModule`.

Comment (monxalo) [reply src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt:23]: Nevermind, these serializers are unique, we can't register another like: `PageResponse.serializer(Parts.serializer))`. Will revert this.

**Comment by monxalo** [comment]
What do you guys think about using an extension like so:
```kotlin
suspend inline fun <reified T> ApplicationCall.respondPaged(message: PageResponse<T>) {
    response.pipeline.execute(this, JsonSettings.toJson(message))
}
```
With this the call would be cleaner instead of this:
```kotlin
call.respond(
    JsonSettings.toJson(
         PageResponse.from(
             pageRequest = pageRequest,
             totalElements = totalElements,
             data = data,
             path = call.request.path()
         )
    )
)
```
After:
```kotlin
call.respondPaged(
    PageResponse.from(
       pageRequest = pageRequest,
       totalElements = totalElements,
       data = data,
       path = call.request.path()
   )
)
```

The main reason for requiring the `JsonSettings.toJson(message)` is because of signature of the method of Ktor's
```kotlin
ApplicationCall.respond(message: Any) 
```
Because this methods accepts `Any` type erasure occurs and the generic type definitions is lost.

**Review by felix19350 — DISMISSED** [review]

**Comment by pedrohh** [comment]
Just so you know `Content-Type` is being returned as **text/plain** instead of **application/json** in the **/cards** endpoint, probably because we are returning a string from the `JsonSettings.toJson(PageRequest<T>)` call.

**Comment by monxalo** [comment]
@pedrohh You're right, so even without my suggestion this still happens when calling:
```kotlin
JsonSettings.toJson(
         PageResponse.from(
             pageRequest = pageRequest,
             totalElements = totalElements,
             data = data,
             path = call.request.path()
         )
    )
```

**Review by monxalo — APPROVED** [review]
